### ISSUE TITLE
UX: improve log formatting & info (on refresh & show)

### ISSUE THREAD

**Issue by DavidGOrtega** [iterative/terraform-provider-iterative#392]
Log what we store in logs and events within the terraform status using the terraform CLI during refresh using [logrus](https://github.com/Sirupsen/logrus). This way would be easier to visualise the logs within the Terraform CLI.

This also applies to current existing logs in creation and destroy.

According to @dmpetrov feedback ideally:

- [ ] Instance lifecycle
     - Creation with info about price
     - Termination
- [ ] Data syncing
- [ ] Script
     - [ ] Setup environment
     - [ ] Script logs
     - [ ] Status
       - Started/Terminated
       - Success/Failed

**Comment by 0x2b3bfa0** [comment]
> _Log what we store in logs and events within the terraform status using the terraform CLI during refresh using [logrus](https://github.com/Sirupsen/logrus). This way would be easier to visualise the logs within the Terraform CLI._

Sounds interesting. Can you please add a bit more of context?

* Why and how should we use a logging library like [logrus](https://pkg.go.dev/github.com/sirupsen/logrus)?
    * Terraform providers can't print to the command–line tool [standard output/error](https://en.wikipedia.org/wiki/Standard_streams) at will; see [hashicorp/terraform-plugin-sdk#145](https://github.com/hashicorp/terraform-plugin-sdk/issues/145) for more information.[^1]
    * Logs can be sent through [hooks](https://github.com/sirupsen/logrus/tree/master/hooks) ([#393](https://github.com/iterative/terraform-provider-iterative/issues/393)) instead, but it might be as (in)convenient for the end user as running `terraform show` or saving logs locally by other means.
* Should we try to alter the behavior of `terraform refresh` to display additional information?

It looks like we're starting to find out that Terraform wasn't meant as a command–line application prototyping toolkit, but rather as a specialized tool to manage persistent (i.e. non ephemeral) infrastructure.

[^1]: Users can set the [`TF_LOG`](https://www.terraform.io/cli/config/environment-variables#tf_log) environment variable to collect provider logs in real time, but relying on this behavior is unrealistic from the user experience standpoint.

**Comment by DavidGOrtega** [comment]
> Logs can be sent through [hooks](https://github.com/sirupsen/logrus/tree/master/hooks) (https://github.com/iterative/terraform-provider-iterative/issues/393) instead

Exactly we could control a bit where or how to see the logs aside the infamous `TF_LOG` like syslog out-of-the-box

> from the client side and requiring users to run terraform refresh periodically
In reality this task is much more looking at the future. At some point we should have all the logic control within go. WE need to create that code in go that handles the task and do the data syncing in go... a proper executor engine and no the current solution that allow us to have more control on this and other things. But at least I can see the logs on refresh without having to read uncomfortable json in the command line

> Should we try to alter the behavior of terraform refresh to display additional information
Not that much than the having logs if the setup `TF_LOG` or thanks to logrus we can control a bit the logs.

**Comment by 0x2b3bfa0** [comment]
> _In reality this task is much more looking at the future. At some point we should have all the logic control within go. WE need to create that code in go that handles the task and do the data syncing in go... a proper executor engine and no the current solution that allow us to have more control on this and other things._

Utterly off–topic, but I agree.

**Comment by DavidGOrtega** [comment]
In any case choosing Terraform or not is not to mi mind our biggest mistake. I do not really like to have the control flow spread in many paradigms... go, bash, etc..

**Comment by 0x2b3bfa0** [comment]
> _But at least I can see the logs on refresh without having to read uncomfortable json in the command line_

Do you mean that the sole fact of running `terraform refresh` would print logs to the standard output? 🤔 It wuld be clearer if you added some examples.

**Comment by 0x2b3bfa0** [comment]
> _But at least I can see the logs on refresh without having to read uncomfortable json in the command line_

Do you mean that the sole fact of running `terraform refresh` would print logs to the standard output? 🤔 It wuld be clearer if you added some examples.

**Comment by DavidGOrtega** [comment]
> Do you mean that the sole fact of running terraform refresh would print logs to the standard output? 🤔 It would be clearer if you added some examples.

With export TF_LOG im able to see the logs using logrus we could also output to something more friendly, and when we have the executor in place we could see real time logs where ever we wanted

**Comment by DavidGOrtega** [comment]
At least the MVP is:
I refresh and I can see whats going on easier than having to grep or cat a json file, even if I have to enable TF_LOG. Additionally we could use syslog via logrus and only that opens an entire world of centralised logging

**Comment by DavidGOrtega** [comment]
The last sentence is more from a UX perspective. In any case the task is discern an implement how we can bring the information that the users are demanding. Regarding to the events and task information. Even if at the last we are going to use pure JSON.

**Comment by 0x2b3bfa0** [comment]
> _With export TF_LOG im able to see the logs using logrus_

Of course, but we can't ask users to manually set `TF_LOG` to use our provider. 🙃 Also, enabling that logging mode will also add noise from Terraform debug messages.

> _we could also output to something more friendly_

For example? A file?

> _and when we have the executor in place we could see real time logs where ever we wanted_

How?

**Comment by 0x2b3bfa0** [comment]
> _I refresh and I can see whats going on easier than having to grep or cat a json file_

Do you mean that `terraform refresh` would output logs directly like this?

```hcl
resource "iterative_task" "example" {
  script    = <<-END
    #!/bin/sh
    echo "Hello world from the cloud!"
  END
}
```

```console
$ terraform refresh
iterative_task.example: Refreshing state... [id=tpi-example-a1b2c3-d4e5f6]
[INFO] Hello world from the cloud!
```

> _even if I have to enable TF_LOG._

Again, this is a line we shouldn't probably cross. Asking users to enable this feature during the normal operation of the program is rather cumbersome and inelegant.

**Comment by 0x2b3bfa0** [comment]
> _Additionally we could use syslog via logrus and only that opens an entire world of centralised logging_

Centralised logging from the client side sounds like the worst of both worlds. 🤔 Still, “centralised logging” in general sounds like an interesting topic to explore after the public release.

**Comment by 0x2b3bfa0** [comment]
> _The last sentence is more from a UX perspective._

Which sentence?

> _In any case the task [1] is discern an implement how we can bring the information that the users are demanding. Regarding to the events and task [2] information.

If “the task [1]” means “our job / this issue” and “task [2]” means `task`, I enthusiastically agree.

> _Even if at the last we are going to use pure JSON._

Like in https://github.com/iterative/terraform-provider-iterative/issues/389#issuecomment-1036090850? 🤔 Doesn't sound especially user friendly.

**Comment by DavidGOrtega** [comment]
> For example? A file?

Syslog i.e. being able to redirect syslog to many decentralised logging solutions.

> Again, this is a line we shouldn't probably cross.

Ehrr... why not? If we just only have this way to enable logs... at least we have somehow and they are able to see it properly

> If “the task [1]” means “this issue” and “task [2]” means task, I enthusiastically agree.

correct

**Comment by 0x2b3bfa0** [comment]
> _Ehrr... why not? If we just only have this way to enable logs... at least we have somehow and they are able to see it properly_

Terraform explicitly forbids providers from interacting with the terminal standard input/output/error for good reasons. It's an infrastructure management tool, not a command–line interface toolkit.

* https://github.com/hashicorp/terraform/issues/17267#issuecomment-1036692773
* https://github.com/hashicorp/terraform-plugin-sdk/issues/145

Asking users to set [`TF_LOG=INFO`](https://www.terraform.io/internals/debugging#debugging-terraform) before using our provider would be an obnoxious design choice for the following reasons:

1. **Noise:** this debug mode doesn't just print the logs we want, but every log from Terraform Core and every provider; running `terraform refresh` with the `INFO` log level already prints 38 lines by default.
2. **Order:** plan/apply tasks run in parallel and log lines aren't guaranteed to be printed contiguously, and output may contain interleaved logs from different providers and resources; this doesn't seem especially beneficial for user experience.
3. **Format:** providers don't have any control over the log line prefix, which adds 30+ characters with a worthless local timestamp to every line.
4. **Convenience:** requiring users to enable this mode as part of the default experience is just [chaotic evil](https://xkcd.com/2408).

**Comment by 0x2b3bfa0** [comment]
> _Syslog i.e. being able to redirect syslog to many decentralised logging solutions._

It's still unclear to me what's the use for `syslog` in this scenario:

1. Legacy [`syslog`](https://en.wikipedia.org/wiki/Syslog) was replaced by [`systemd-journald`](https://www.freedesktop.org/software/systemd/man/systemd-journald.service.html) long ago. Modern distributions don't even support `syslog` network logging out of the box. https://github.com/systemd/systemd/issues/15200.

* Having **our provider** write **cloud machine task logs** locally to **the system where it runs** is similarly [chaotic evil](https://xkcd.com/2408). Logs in the journal are interleaved with other system logs, thus hard to parse; and providers aren't expected to have this kind of side effect anyway.

* Having **our provider** write **cloud machine task logs** to **a remote service** — if that's what you propose — doesn't seem especially useful, because it would require users to run `terraform refresh` periodically.

* Having **cloud machines** write task logs to [**appropriate cloud services**](https://github.com/iterative/terraform-provider-iterative/issues/393#issuecomment-1036108767) periodically, instead/apart of writing them to object storage — current approach — could be a nice addition, but doesn't solve any of our immediate issues.

**Comment by DavidGOrtega** [comment]
> Having our provider write cloud machine task logs locally to the system where it runs is similarly chaotic. Logs in the journal are interleaved with other system logs, thus hard to parse; and providers aren't expected to have this kind of side effect anyway.

When I read this I have the impression that you never had to deal with logs

**Comment by DavidGOrtega** [comment]
> Asking users to set [TF_LOG=INFO](https://www.terraform.io/internals/debugging#debugging-terraform) before using our provider would be a controversial design choice for the following reasons

The whole team agreed on using TF_LOG
